stan/math/rev/core/var.hpp

```diff
@@ -25,36 +23,49 @@ static void grad(vari* vi);
  * an arena-based memory manager scoped to a single gradient
  * calculation.
  *
- * A var is constructed with a double and used like any
+ * A var is constructed with a type `T` and used like any
  * other scalar.  Arithmetical functions like negation, addition,
  * and subtraction, as well as a range of mathematical functions
  * like exponentiation and powers are overridden to operate on
  * var values objects.
+ * @tparam T An Arithmetic type.
  */
-class var {
+template <typename T, typename = void>
+class var_value {};
+
+template <typename T>
+class var_value<T, require_floating_point_t<T>> {
  public:
-  // FIXME: doc what this is for
-  using Scalar = double;
+  // The internal value_type is always a floating point type
+  using value_type = std::decay_t<T>;
```

Comment:
So `value_type` here is what you get when you'd do `value_of`. Is that logic extendable? Is this how we define `value_type_t` too?

Reply:
I think keeping those separate is a good idea. I just don't really know the benefit of mooshing them together in a meaningful way

Reply:
Well there are three different things vaguely called values (`value_of`, `value_type` in classes, and `value_type_t`). In two of those cases I *think* things are primitive types, though I could easily see an argument for a `var` being a value.

Reply:
It's really confusing to have them conceptually separate, is my point.

Reply:
`value_type_t` is just an alias for `value_type` ala

```cpp
template <typename T>
using value_type_t = typename value_type<T>::type;
```
it just stops us having to write the `typename` and `::type` everywhere.

`value_of` won't always be the exact type. Like this one

```cpp
template <typename EigMat, require_eigen_t<EigMat>* = nullptr,
          require_not_vt_double_or_int<EigMat>* = nullptr>
inline auto value_of(const EigMat& M) {
  return M.unaryExpr([](const auto& scal) { return value_of(scal); });
}
``

For `Eigen::Matrix<var>` this returns back an Eigen::Matrix<double>, but `value_type_t<Eigen::Matrix<var>>` would return back `var`.

Reply:
> value_of won't always be the exact type

Gaaah. You're right. This makes sense too.

> `value_type_t<Eigen::Matrix<var>>` would return back `var`

I'll back off before we spawn another type traits thread.

Reply:
> For Eigen::Matrix<var> this returns back an Eigen::Matrix

Actually there is no .eval() here anymore - this is returning an expression!

Reply:
Yep! I was simplifying it a bit but `value_of` returns a different thing then `value_type`